gdnative-core/src/object.rs

```diff
@@ -248,530 +377,629 @@ impl<T: GodotObject + ManuallyManaged> Ptr<T> {
     ///
     /// [thread-safety]: https://docs.godotengine.org/en/stable/tutorials/threads/thread_safe_apis.html
     #[inline(always)]
-    pub unsafe fn assume_safe<'a>(self) -> &'a T {
-        debug_assert!(
-            self.is_instance_sane(),
-            "assume_safe called on an invalid pointer"
-        );
-        self.assume_safe_unchecked::<'a>()
+    pub unsafe fn assume_safe<'a>(&self) -> TRef<'a, T, Shared> {
+        T::RefKind::impl_assume_safe(self)
+    }
+
+    /// Assume that `self` is the unique reference to the underlying object.
+    ///
+    /// This is guaranteed to be a no-op at runtime if `debug_assertions` is disabled. Runtime
+    /// sanity checks may be added in debug builds to help catch bugs.
+    ///
+    /// # Safety
+    ///
+    /// Calling `assume_unique` when `self` isn't the unique reference is instant undefined
+    /// behavior. This is a much stronger assumption than `assume_safe` and should be used with
+    /// care.
+    #[inline(always)]
+    pub unsafe fn assume_unique(self) -> Ref<T, Unique> {
+        T::RefKind::impl_assume_unique(self)
+    }
+}
+
+impl<T: GodotObject<RefKind = ManuallyManaged>> Ref<T, Shared> {
+    /// Returns `true` if the pointer currently points to a valid object of the correct type.
+    /// **This does NOT guarantee that it's safe to use this pointer.**
+    ///
+    /// # Safety
+    ///
+    /// This thread must have exclusive access to the object during the call.
+    #[inline]
+    pub unsafe fn is_instance_sane(self) -> bool {
```

Comment:
Wouldn't it be more appropriate to use `&self` here? I know that all `Shared` references implement `Copy` but it still feels a bit odd.

Reply:
Changed to `&self`.